public/lib/uploadHttpConcurrentProgress.js

```diff
@@ -0,0 +1,277 @@
+                    }
+                }
+
+
+                singleMovingAverage = singleMovingAverage / lastElem;
+                totalMovingAverage = totalMovingAverage + singleMovingAverage;
+            }
+
+        }
+        this.clientCallbackProgress(totalMovingAverage);
+        this._finalResults.push(totalMovingAverage);
+    };
+
+    /**
+     * onProgress method
+     */
+    uploadHttpConcurrentProgress.prototype.onTestProgress = function (result) { // jshint ignore:line
+
+        if (!this._running) {
+            return;
+        }
+
+        //update progress count
+        this._progressCount++;
+
+        this.testResults.push(result.bandwidth);
+        //populate array
+        this._progressResults['arrayProgressResults' + result.id].push(result.bandwidth);
+        //calculate moving average
+        if (this._progressCount % this.movingAverage === 0) {
+            var example = this.testResults.splice(0, 11);
+
+            var singleMovingAverage = 0;
+            for (var i = 0; i < this.movingAverage; i++) {
+                singleMovingAverage += example[i];
+            }
+
+            if (isFinite(singleMovingAverage)) {
+                var totalMovingAverage = singleMovingAverage/this.movingAverage;
+                this.clientCallbackProgress(totalMovingAverage);
+                this._finalResults.push(totalMovingAverage);
+            }
+
+        }
+    };
+
+    /**
+     * Start the test
+     */
+    uploadHttpConcurrentProgress.prototype.start = function () {
+        var request;
+
+        if (!this._running) {
+            return;
+        }
+
+        if (this.type === 'GET') {
+            for (var g = 1; g <= this.concurrentRuns; g++) {
+                this._testIndex++;
+
+                this['arrayResults' + this._testIndex] = [];
+                this._progressResults['arrayProgressResults' + this._testIndex] = [];
+                request = new window.xmlHttpRequest('POST', this.url, this.timeout, this.onTestComplete.bind(this), this.onTestProgress.bind(this),
+                    this.onTestAbort.bind(this), this.onTestTimeout.bind(this), this.onTestError.bind(this));
+                this._activeTests.push({
+                    xhr: request,
+                    testRun: this._testIndex
+                });
+                request.start(0, this._testIndex);
+
+            }
+        }
+        else {
+            for (var p = 1; p <= this.concurrentRuns; p++) {
+                this._testIndex++;
+                this['arrayResults' + this._testIndex] = [];
+                this._progressResults['arrayProgressResults' + this._testIndex] = [];
+                request = new window.xmlHttpRequest('POST', this.url, this.timeout, this.onTestComplete.bind(this), this.onTestProgress.bind(this),
+                    this.onTestAbort.bind(this), this.onTestTimeout.bind(this), this.onTestError.bind(this));
+                this._activeTests.push({
```

Comment:
If the uploads bandwidth test creates multiple http post requests then we would want to create the payload once in this file and pass in the payload to the xmlhttprequest object instead or recreating for each request